### ISSUE TITLE
Configuration Profiles offered as Conditional Managed Installs behave as regular Managed Installs

### ISSUE THREAD

**Issue by gerritdewitt** [munki/munki#628]
Hello,

Hopefully I didn't miss any known limitations in the documentation, so please let me know if I did.  Here's the issue:

Suppose you have a manifest with a conditional item like the one below.  In at least Munki 2.7.0.2753, the config profile offered as a conditional managed installation will be installed (subject to standard tests for a managed install) regardless of the condition's evaluation.  In other words, if you have a new config profile to be installed selectively and automatically, it just gets installed on all systems with the condition as if the condition evaluated true.

<pre>
&lt;key&gt;conditional_items&lt;/key&gt;
&lt;array&gt;&lt;dict&gt;
&lt;key&gt;condition&lt;/key&gt;
   &lt;string&gt;some_condition == TRUE&lt;/string&gt;
&lt;key&gt;managed_installs&lt;/key&gt;
   &lt;array&gt;
      &lt;string&gt;Some_Profile-2016.07.mobileconfig&lt;/string&gt;
   &lt;/array&gt;
&lt;/dict&gt;&lt;/array&gt;
</pre>

This can be problematic in cases where you're using a conditional item as I believe one would typically; as examples:

1.  A config profile to bind systems to Active Directory (com.apple.DirectoryService.managed payload) would only need to be installed if a condition determined the system was on the appropriate network but unbound.  Same for other directory domains.  Imagine having a condition that would return a "needs binding" state; you'd want to do a conditional managed install for the profile.  (I am aware that a config profile isn't the only way to do this, of course.)

2.  You may want to only offer AD Certificate or 802.1X configuration profiles if the AD state indicates the system is bound.  This is especially useful if the binding happened via dsconfigad or the system was bound prior to joining it to Munki; a condition to check for AD connectivity would evaluate true and the profile(s) would be installed based on that.

--Gerrit

**Comment by gregneagle** [comment]
Please post output from `managedsoftwareupdate -vvv` demonstrating this behavior.
As described this looks like an improperly evaluated condition and not at all connected to what you are installing.

**Comment by erikng** [comment]
To add to this, I have configuration profiles with conditions and they are working correctly.

**Comment by gerritdewitt** [comment]
Hi Greg,

Many thanks for the super quick follow up! Sorry I wasn't able to reciprocate. I can provide a log, but will need to redact it a bit.  I am not finished doing that just yet.

However, in looking at this, I have a better idea of what's going on:

- I can confirm the predicate is being evaluated as expected. In this case, we use a condition to report on the status of an AD bind. The condition is properly returning that the system is on the network and bound. (The condition reports three states: off network, on network and unbound, on network and bound. We look for on network and unbound to conditionally do a managed install for a profile with one com.apple.DirectoryService.managedpayload for [re] binding the system to AD.)

- We also have a config profile with com.apple.AdCertificate and com.apple.wifi.managed payloads for setting up EAP-TLS Wi-Fi. The pkginfo for this mobileconfig requires the previous AD binding profile.

- We periodically modify our config profiles as details change, and I think what's happened here is the Wi-Fi profile requires the AD binding one, which changed. So Munki compares the file hash of the config profile and sees it as something different even though we make the identifiers the same. I can consistently get the AD binding profile to reinstall if I clear the ConfigProfileDara.plist.

- I am going to try switching the Wi-Fi config profile to a conditionally offered one without the required dependency on the AD binding profile in its pkginfo. I think that will fix my issue.

Sent from my iPhone

On Jun 29, 2016, at 11:59 AM, Greg Neagle <notifications@github.com<mailto:notifications@github.com>> wrote:


Please post output from managedsoftwareupdate -vvv demonstrating this behavior.
As described this looks like an improperly evaluated condition and not at all connected to what you are installing.

-

**Comment by gerritdewitt** [comment]
Hello everyone,

I've made a config change removing the "requires" key in the pkginfo for our Wi-Fi config (see my previous post).  So this issue seems to be a false alarm.  Thank you to everyone for helping me resolve this.

In troubleshooting, I've learned that Munki determines if a configuration profile is installed by its profile identifier, install data, and file hash.  Instead of file hash, could we use the profile's UUID?  I'm thinking about cases where we have automations that carefully craft the payload contents, but for which we can keep the profile identifier and UUID keys the same.  This may be a very narrow use case, though.  In other words, even if we change the profile's content, we could consider it already installed if the top UUID and top identifier were kept the same.  Just thinking out loud here.

We can close this issue.  Thank you once again for your help!
